p2p/remote.go

```diff
@@ -0,0 +1,98 @@
+package p2p
+
+import (
+	"context"
+
+	manet "gx/ipfs/QmV6FjemM1K8oXjrvuq3wuVWWoU2TLDPmNnKrxHzY3v6Ai/go-multiaddr-net"
+	ma "gx/ipfs/QmYmsdtJ3HsodkePE3eU3TsCaP2YvPZJ4LoXnNkDE5Tpt7/go-multiaddr"
+	net "gx/ipfs/QmZNJyx9GGCX4GeuHnLB8fxaxMLs4MjTjHokxfQcCd6Nve/go-libp2p-net"
+	protocol "gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
+)
+
+var maPrefix = "/" + ma.ProtocolWithCode(ma.P_IPFS).Name + "/"
+
+// remoteListener accepts libp2p streams and proxies them to a manet host
+type remoteListener struct {
+	p2p *P2P
+
+	// Application proto identifier.
+	proto protocol.ID
+
+	// Address to proxy the incoming connections to
+	addr ma.Multiaddr
+}
+
+// ForwardRemote creates new p2p listener
+func (p2p *P2P) ForwardRemote(ctx context.Context, proto protocol.ID, addr ma.Multiaddr) (Listener, error) {
+	listener := &remoteListener{
+		p2p: p2p,
+
+		proto: proto,
+		addr:  addr,
+	}
+
+	if err := p2p.ListenersP2P.Register(listener); err != nil {
+		return nil, err
+	}
+
+	return listener, nil
+}
+
+func (l *remoteListener) handleStream(remote net.Stream) {
+	local, err := manet.Dial(l.addr)
+	if err != nil {
+		remote.Reset()
+		return
+	}
+
+	peer := remote.Conn().RemotePeer()
+
+	peerMa, err := ma.NewMultiaddr(maPrefix + peer.Pretty())
+	if err != nil {
+		remote.Reset()
+		return
+	}
+
+	stream := &Stream{
+		Protocol: l.proto,
+
+		OriginAddr: peerMa,
+		TargetAddr: l.addr,
+		peer:       peer,
+
+		Local:  local,
+		Remote: remote,
+
+		Registry: l.p2p.Streams,
+	}
+
+	l.p2p.Streams.Register(stream)
+}
+
+func (l *remoteListener) Protocol() protocol.ID {
+	return l.proto
+}
+
+func (l *remoteListener) ListenAddress() ma.Multiaddr {
+	addr, err := ma.NewMultiaddr(maPrefix + l.p2p.identity.Pretty())
+	if err != nil {
+		panic(err)
+	}
+	return addr
+}
+
+func (l *remoteListener) TargetAddress() ma.Multiaddr {
+	return l.addr
+}
+
+func (l *remoteListener) Close() error {
+	_, err := l.p2p.ListenersP2P.Deregister(string(l.proto))
```

Comment:
Now that we're just registering by *protocol*, this is going to have that close race condition we discussed. That is, if I try to close a listener forwarding to a *specific* local address twice while someone else is trying to *open* a listener (to a different local address), I can close their listener.

I believe the correct fix is to just do *everything* inside the listener manager. That is, no listener handles, etc. Just add a `Close(...)` method to the listener's object that takes the close "constraints" (protocol, localAddr, remoteAddr).

Does that make sense? This is what I meant by reducing some of the abstractions.